cli/popper/runner_slurm.py

```diff
@@ -0,0 +1,127 @@
+import os
+from subprocess import PIPE, Popen, STDOUT, SubprocessError
+
+import docker
+
+from popper import utils as pu
+from popper import scm
+from popper.cli import log as log
+from popper.runner import StepRunner as StepRunner
+from popper.runner_host import DockerRunner as HostDockerRunner
+from popper.runner_host import HostRunner
+
+
+class SlurmRunner(StepRunner):
+
+    spawned_processes = set()
+
+    def __init__(self, config):
+        super(SlurmRunner, self).__init__(config)
+
+    def __exit__(self, exc_type, exc, traceback):
+        SlurmRunner.spawned_processes = set()
+
+    def exec_srun_cmd(self, cmd):
+        try:
+            cmd.insert(0, 'srun')
+            with Popen(cmd, stdout=PIPE, stderr=STDOUT,
+                       universal_newlines=True, preexec_fn=os.setsid,
+                       cwd=self.config.workspace_dir) as p:
+                SlurmRunner.spawned_processes.add(p)
+
+                log.debug('Reading process output')
+
+                for line in iter(p.stdout.readline, ''):
+                    line_decoded = pu.decode(line)
+                    log.step_info(line_decoded[:-1])
+
+                p.wait()
+                ecode = p.poll()
+                SlurmRunner.spawned_processes.remove(p)
+
+            log.debug(f'Code returned by process: {ecode}')
+
+        except SubprocessError as ex:
+            ecode = ex.returncode
+            log.step_info(f"Command '{cmd[0]}' failed with: {ex}")
+        except Exception as ex:
+            ecode = 1
+            log.step_info(f"Command raised non-SubprocessError error: {ex}")
+
+        return ecode
+
+    def stop_srun_cmd(self):
+        for p in SlurmRunner.spawned_processes:
+            log.info(f'Stopping proces {p.pid}')
+            p.kill()
+
+
+class DockerRunner(SlurmRunner, HostDockerRunner):
+    """Runs steps in docker."""
+
+    def __init__(self, config):
+        super(DockerRunner, self).__init__(config)
+
+    def run(self, step):
```

Comment:
we can refactor this function so that we avoid repeating code from `runner_host.DockerRunner`. Maybe create utility function in `runner_host.DockerRunner` that can be reused here?

Reply:
I did some refactoring. Please review.